implement backslash as a prefix for reserved keywords (#93)
FAIRmat-NFDI/nyaml: docs/learn/yaml-notation.md

Reviewer: ```suggestion
    <group name="transmission_correction" type="NXcalibration" optional="true">
```

Author: Suggestion looks weird, I just corrected it myself. Thanks

Reviewer: ```suggestion
          \doc: |
```

Author: Sugggestion was ill-formatted, I just changed it myself.

Reviewer: ```suggestion
      \doc:
```

Author: Suggestion did not work, I changed it myself.

## Special keywords

### Reserved keywords and escape characters

> `nyaml` uses a set of reserved keys, each written with a `\` prefix, that map to NXDL concept properties. The `\` prefix activates the corresponding feature; writing the same key *without* the `\` prefix makes it a plain concept name (field, group, or attribute) instead.

Reviewer: ```suggestion
`nyaml` uses a set of reserved keys, each written with a `\` prefix, that map to NXDL concept properties. The `\` prefix activates the corresponding feature; writing the same key *without* the `\` prefix makes it a plain concept name (field or group) instead, and  `\@` for attributes. 
```

Reviewer 2: ", and  `\@` ..." > "instead. Use the prefix `\@` for defining attributes.

**Reserved keywords**

| Keyword | Where it applies |
|---------|-----------------|
| `\doc` | doc string for any concept |

> | `\unit` | unit category for a field |

Reviewer 2: ```suggestion
| `\unit` | NeXus unit category for a field |
```

| `\enumeration` | enumeration values for a field or attribute |

> | `\nameType` | name-matching rule (`specified`, `any`, `partial`) |

Reviewer 2: ```suggestion
| `\nameType` | enumeration for the name-matching rule (`specified`, `any`, `partial`) |
```

### `dimensions`

> Describes the shape of a multidimensional data array. Using the "dimensions" keyword constraints the instance data of a field to an array with at least one dimension. The `\rank` key gives the number of dimensions. Each dimension is specified with `\dim` using either the full array-of-arrays form or the shorter tuple form. Use symbolic names from the root `\symbols` block rather than hardcoded integers.

Reviewer 2: remove the word "multidimensional" in the first sentence

Author: done

=== "YAML (open enum)"
    ```yaml

>     enum_with_open_enum:

Reviewer: ```suggestion
    \enum_with_open_enum:
```

Reviewer 2: No, the initial version is correct, a concept called "enum_with_open_enum"

Author: No, this is a field name where we don't want to have the slash.

      \enumeration:
        \open: true

>         items: [NXmpes]

Reviewer: ```suggestion
        \items: [NXmpes]
```

Reviewer 2: using "items" here at all is inconsistent with L395

Author: I had missed that keyword `\items` so far, I have enabled it now

=== "YAML (open enum with doc)"
    ```yaml
    enum_with_open_and_vector_items:
      \enumeration:
        \open: true
        '[0, 1, 0]':

>           doc: |

Reviewer: ```suggestion
          \doc: |
```

### `xref`

> The `\xref` keyword (only valid inside `doc`) links a concept to an entry in an external ontology or standard such as ISO. It is YAML-only: `nyaml` converts it to the corresponding textual representation inside the XML `<doc>` element.

Reviewer 2: ```suggestion
The `\xref` keyword (only valid inside `\doc`) links a concept to an entry in an external ontology or standard such as ISO. It is YAML-only: `nyaml` converts it to the corresponding textual representation inside the XML `<doc>` element.
```

      </doc>
    </group>
    ```

---

## NXDL coverage

`nyaml` covers all NXDL concepts that are relevant to writing modern NeXus definitions. The table below maps each major NXDL XSD type and its attributes to the corresponding nyaml notation.

| NXDL XSD concept | Attributes / children | nyaml notation |
|---|---|---|
| `definition` | `category` | `\category` |
| | `type` | `\type` |
| | `name`, `extends` | `NXname(NXbase):` key |
| | `restricts` | `\restricts` |
| | `deprecated` | `\deprecated` |
| | `ignoreExtraGroups` | `\ignoreExtraGroups` |
| | `ignoreExtraFields` | `\ignoreExtraFields` |
| | `ignoreExtraAttributes` | `\ignoreExtraAttributes` |
| `group` | `name`, `type` | `name(NXbaseclass):` key |
| | `minOccurs`, `maxOccurs` | `\minOccurs`, `\maxOccurs` |

> | | `recommended`, `optional` | `\exists: recommended / optional` |

Reviewer 2: ```suggestion
| | `recommended`, `optional` | `\exists: recommended`, `\exists: optional` |
```

| | `nameType` | `\nameType` |
| | `deprecated` | `\deprecated` |
| `field` | `name` | plain key, e.g. `fieldname:` |
| | `type` | `fieldname(NX_TYPE):` |
| | `units` | `\unit` |
| | `minOccurs`, `maxOccurs` | `\minOccurs`, `\maxOccurs` |

> | | `recommended`, `optional` | `\exists: recommended / optional` |

Reviewer 2: ```suggestion
| | `recommended`, `optional` | `\exists: recommended`, `\exists: optional` |
```

| | `nameType` | `\nameType` |
| | `deprecated` | `\deprecated` |
| | `dimensions` child | `\dimensions` with `\rank`, `\dim` |
| | `enumeration` child | `\enumeration` |
| | `attribute` child | `\@name:` |
| `attribute` | `name`, `type` | `\@name(NX_TYPE):` |

> | | `recommended`, `optional` | `\exists: recommended / optional` |

Reviewer 2: ```suggestion
| | `recommended`, `optional` | `\exists: recommended`, `\exists: optional` |
```

| | `nameType` | `\nameType` |
| | `enumeration` child | `\enumeration` |
| `link` | `name` | `name(link):` key |
| | `target` | `\target` |
| | `napimount` | `\napimount` |

> | `enumeration` | `open` | `\open: true` |

Reviewer 2: Is using `\open: false` supported or discouraged?

Author: `\open: false` is discouraged; it's redundant and will be dropped on round-trip.

Behaviour: 
- Forward: it passes the value through literally, so `\open: false` produces `open="false"` in the XML.
- Backward only emits `\open: true` when the XML attribute `open="true"``— it never emits `\open: false`.

So `\open: false` round-trips as: YAML `\open: false` → XML `open="false"` → YAML (no `\open` key) . The `false`  value is silently dropped.

The main reason is that since closed enumeration are the default, we don't need to re-write them all the time (it's like `NX_CHAR` in that sense). We basically only need to react to `open="true"` in the XML.

Author: I added an explanation to the docs and some test cases in `tests/data/enumerations.yaml`

Reviewer 2: That's good, thanks for clarifying it here and in the docs as well.
